traversal/producer/GraphProducer.java

```diff
@@ -46,91 +47,102 @@
     private final List<Identifier.Variable.Name> filter;
     private final ResourceIterator<? extends Vertex<?, ?>> start;
     private final ConcurrentHashMap.KeySetView<VertexMap, Boolean> produced;
+    private final Map<ResourceIterator<VertexMap>, CompletableFuture<Void>> runningJobs;
     private final AtomicBoolean isDone;
-    private final Map<ResourceIterator<VertexMap>, CompletableFuture<Void>> iteratorJobs;
-    private final Map<ResourceIterator<VertexMap>, Integer> iteratorRequested;
+    private boolean isInitialised;
 
-    public GraphProducer(GraphManager graphMgr, GraphProcedure procedure, Traversal.Parameters params, List<Identifier.Variable.Name> filter, int parallelisation) {
+    public GraphProducer(GraphManager graphMgr, GraphProcedure procedure, Traversal.Parameters params,
+                         List<Identifier.Variable.Name> filter, int parallelisation) {
         assert parallelisation > 0;
         this.graphMgr = graphMgr;
         this.procedure = procedure;
         this.params = params;
         this.filter = filter;
         this.parallelisation = parallelisation;
-        this.isDone = new AtomicBoolean(false);
+        this.runningJobs = new HashMap<>();
         this.produced = ConcurrentHashMap.newKeySet();
         this.start = procedure.startVertex().iterator(graphMgr, params);
-        this.iteratorJobs = new HashMap<>();
-        this.iteratorRequested = new HashMap<>();
+        this.isDone = new AtomicBoolean(false);
+        this.isInitialised = false;
     }
 
     @Override
-    public synchronized void produce(Queue<VertexMap> queue, int count) {
-        if (iteratorRequested.size() < parallelisation) {
-            for (int i = iteratorRequested.size(); i < parallelisation && start.hasNext(); i++) {
-                ResourceIterator<VertexMap> iterator =
-                        new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
-                iteratorRequested.put(iterator, 0);
-            }
+    public synchronized void produce(Producer.Queue<VertexMap> queue, int request) {
+        if (isDone.get()) return;
```

Comment:
Is it possible that `isDone` is set to `true` after this check? Is this OK?

Reply:
It is possible. But I think it's okay?